2018/001-containers-tagging-strategy.md

```diff
@@ -0,0 +1,189 @@
+`docker images` output):
+
+```
+REPOSITORY           TAG                    IMAGE ID
+caasp/mariadb        10.2                   1ade07d13d13
+caasp/mariadb        latest                 1ade07d13d13
+caasp/mariadb        10.2.15-3.2            1ade07d13d13
+```
+
+Note that the build ID is being used as the release number. Build ID are
+the build numbers currently included in the resulting files of KIWI build
+within the build service.
+
+All three tags pointing to the same image. Then if a new update due to a new
+mariadb release happens adding the new image could look like that:
+
+```
+REPOSITORY           TAG                    IMAGE ID
+caasp/mariadb        10.2                   70b5d81549ec
+caasp/mariadb        latest                 70b5d81549ec
+caasp/mariadb        10.2.15-3.2            1ade07d13d13
+caasp/mariadb        10.2.15-3.3            70b5d81549ec
+```
+
+Note that old image would be still accessible with the concrete version tight
+to the specific version and release. Again, if a new update happens but this
+time due to a new mariadb version adding a third image into the registry could
+result in:
+
+```
+REPOSITORY           TAG                    IMAGE ID
+caasp/mariadb        10.2                   70b5d81549ec
+caasp/mariadb        latest                 284549eacf84
+caasp/mariadb        10.2.15-3.2            1ade07d13d13
+caasp/mariadb        10.2.15-3.3            70b5d81549ec
+caasp/mariadb        10.3                   284549eacf84
+caasp/mariadb        10.3.1-3.4             284549eacf84
+```
+
+Aging all three image versions are accessible and there is a stable reference
+to get the latest. However this time, since the mariadb version changed, there
+is also version tag to get the latest release of an specific version.
+
+To get to this sort of tagging strategy some modifications are required in our
+workflow and toolchain.
+
+* Need correct versions of KIWI and skopeo in relevant streams (mostly SLE 15)
+  With it the stable tags could be already defined ('latest' and '10.3' of the
+  above example)
+  
+* Need OBS to add/modify tags at build time using the KIWI command call to
+  append the build ID and/or release numbers. Currently the tag can be modified
+  at call time in KIWI with the `--set-container-tag`. Probably this can be
+  used to append release numbers to a tag.
+  
+* Need OBS to push to the registry the image with all the tags included within
+  the built binary.
+
+* SUSE requires some tool to track changes in the registry and be capable to
+  detect when updates are available.
+  
+Tagging an image with the main package version it can be achieved by using the
+`obs-service-replace_using_package_version` service. This is currently being
+used in Head project in IBS & OBS.
+
+One of the advantages of using this strategy is that it does not necessarily
+require a Kubernetes manifest update for each image update. The manifest could
```

Comment:
Without a manifest update, Kubernetes won't restart running containers with the new build of the image - I don't believe we can get away from doing this without hitting the problem you described above.